storage: don't reread value during inline conditional writes

nvb opened the pull request:
This commit removes the call to maybeGetValue when performing an inline conditional write. In such cases, we will have already read the value in the call to mvccGetMetadata, so we don't need to do so again.

I'm not aware of any workloads that are sensitive to the performance of conditional inline writes and I also suspect that the positioning of the Pebble iterator was avoiding some work during the second seek, so this is mainly just intended to be a code simplification.

Epic: None
Release note: None

nvb committed:
storage: add basic test of inline operations to TestMVCCHistories

This commit adds a basic test of inline operations to the
TestMVCCHistories test suite.

Epic: None
Release note: None

nvb committed:
storage: don't reread value during inline conditional writes

This commit removes the call to maybeGetValue when performing an inline
conditional write. In such cases, we will have already read the value in
the call to mvccGetMetadata, so we don't need to do so again.

I'm not aware of any workloads that are sensitive to the performance of
conditional inline writes and I also suspect that the positioning of the
Pebble iterator was avoiding some work during the second seek, so this is
mainly just intended to be a code t simplification.

Epic: None
Release note: None

nvb committed:
storage: delete maybeGetValue

The function is no longer needed. We can inline its one caller and gain
some readability.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109793)
<!-- Reviewable:end -->

itsbilal reviewed (approved):
LGTM

nvb committed:
storage: use putBuffer.newMeta when writing inline value

This commit uses putBuffer.newMeta when writing an inline value, instead
of putBuffer.meta. This avoids a near-bug, because we were consulting
buf.meta.Deleted a few lines down.

Epic: None
Release note: None

nvanbenschoten commented:
TFTR!

bors r=itsbilal
